app/buck2_common/src/http/retries.rs

```diff
@@ -45,15 +45,46 @@ pub trait AsHttpError {
     fn as_http_error(&self) -> Option<&HttpError>;
 }
 
-pub async fn http_retry<Exec, F, T, E>(exec: Exec, mut intervals: Vec<Duration>) -> Result<T, E>
+/// Used to dispatch warnings to the event log in the case in the case of a
+/// retryable HTTP error. This often occurs when some level of exponential
+/// backoff is happening due to a lot of concurrent http_archive() requests for
+/// a single host; see GH-316 and GH-321 for background.
+pub trait DispatchableHttpRetryWarning {
+    /// Fire off a warning. This might go to the console, event log, or
+    /// some other place.
+    fn dispatch(&self, dur: &Duration, retries: usize, url: &str);
+}
+
+/// No-op implementation of DispatchableHttpRetryWarning. This does nothing at
+/// all when a retry occurs; useful for mock tests.
+pub struct NoopDispatchableHttpRetryWarning {}
+
+impl NoopDispatchableHttpRetryWarning {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl DispatchableHttpRetryWarning for NoopDispatchableHttpRetryWarning {
+    fn dispatch(&self, _backoff: &Duration, _retries: usize, _url: &str) {}
+}
+
+pub async fn http_retry<Exec, F, T, E, R>(
+    url: &str,
+    dispatch_retry_warning: R,
```

Comment:
(nit) passing `&R` might make a bit more sense considering that' all the API requires